File: src/app/lib/utils/scoring.dart
/// Skill-weighted scoring for score predictions (improvements.md #8).
///
/// A single prediction is scored against the match's full-time result:
///   • Exact score .............. 5 points
///   • Correct goal difference ... 3 points
///   • Correct result (1X2) ...... 1 point
///   • Otherwise ................. 0 points
///
/// Only the best applicable tier is awarded (they are mutually exclusive in
/// value). "Goal difference" already implies the correct result, and an exact
/// score already implies the correct goal difference.
class Scoring {
  Scoring._();

  static const int exactPoints = 5;
  static const int goalDiffPoints = 3;
  static const int resultPoints = 1;

  /// Points for one prediction ([predA]:[predB]) vs the actual full-time score
  /// ([actualA]:[actualB]).
  static int points(int predA, int predB, int actualA, int actualB) {
    if (predA == actualA && predB == actualB) return exactPoints;
    if (predA - predB == actualA - actualB) return goalDiffPoints;
    if ((predA - predB).sign == (actualA - actualB).sign) return resultPoints;
    return 0;
  }
}

File: src/backend/poller/scoring.py
"""Prediction scoring — mirrors lib/utils/scoring.dart so the poller-computed
leaderboard matches the app's client-side fallback exactly (#8).

  • Exact score .............. 5 points
  • Correct goal difference ... 3 points
  • Correct result (1X2) ...... 1 point
  • Otherwise ................. 0 points
"""

from __future__ import annotations

EXACT_POINTS = 5
GOAL_DIFF_POINTS = 3
RESULT_POINTS = 1


def _sign(n: int) -> int:
    return (n > 0) - (n < 0)


def points(pred_a: int, pred_b: int, actual_a: int, actual_b: int) -> int:
    if pred_a == actual_a and pred_b == actual_b:
        return EXACT_POINTS
    if pred_a - pred_b == actual_a - actual_b:
        return GOAL_DIFF_POINTS
    if _sign(pred_a - pred_b) == _sign(actual_a - actual_b):
        return RESULT_POINTS
    return 0
